URI encode request url

jjcarstens opened the pull request:
~~CLI counterpart of https://github.com/nerves-hub/nerves_hub_web/pull/493~~

~~If a device identifier has a space, it becomes unmanageable from the CLI because that gets sent as a `+` (i.e. `"device 1234" -> device+1234`). So this makes a change to `URI.encode/1` the org name or device identifier in requests so that they are transmitted correctly.~~

EDIT:
If some part of a path has a space or other unsafe URI character, the whole request will fail. This changes to do `URI.encode/1` on any path that is sent for a request to cover multiple bases.
    
Further validation to be done in NervesHubWebCore

jjcarstens commented:
You know, I was kind of thinking the same thing. This came about from @andrew-lewin experimenting with NervesHub and was able to create devices with spaces in the identifier, but then could not remove/correct it.

Maybe I should add an annoying logger warning here if we detect spaces? And change the `create` endpoint to remove support for spaces?

fhunleth commented:
I'm so torn. Part of me wants to be inclusive in what's supported, but the other part knows that there's pain in other places if you use things like spaces, non-Latin characters, emoji, etc.

While it's not required for people to encode their device IDs into barcodes, I think that doing so is extremely useful and we should guide people to label devices. If we're too permissive, people will run into issues with either character set support in barcodes or limitations in the software that makes them. I like the [GS1 spec](https://www.gs1.org/standards/barcodes-epcrfid-id-keys/gs1-general-specifications) when looking at barcodes since it's influential and provides practical subsets on the individual barcode specs. 

The spec is super-long and I'm not sure whether IDs should be more limited, but limiting device IDs to the following set seems good for a start:

![image](https://user-images.githubusercontent.com/64669/59228010-bb9c7080-8ba4-11e9-984f-62842fd97b44.png)
![image](https://user-images.githubusercontent.com/64669/59228020-c6ef9c00-8ba4-11e9-9f4d-8f870479c37f.png)

I'd probably limit it more for products that I work on, but limiting device IDs to the above character set seems like we'll be leading people towards more happier paths than not.

jjcarstens commented:
I like that. Though I think we should do this in `NervesHubWebCore` as part of a validation before doing any DB insert so we can control it as close to source as possible, rather than having to adjust this and other libraries that interact with NervesHub. I'm going to move that portion to an issue in `nerves_hub_web`.

Doing that means this fix is still valid for now as we would need to encode the strings that have supported special characters from this spec (i.e. `<`, `>`, and `%`). We can also do the restriction on spaces there as well which would get reported down in any API request

andrew-lewin commented:
Yeah, we're not planning on using any spaces going forward, just something that happened with a couple devices while learning the ropes that aren't able to be deleted now

mobileoverlord commented:
I just tried to rebase this from main, but it looks like the tests are breaking on it now. :/

jjcarstens commented:
K, just went through the rebase and cleaned this up a bit.

I moved this to just encode every path in `API.request` instead of trying handle each module individually
